### PR TITLE
Fix duplicate heading id

### LINKED ISSUE #857: Hash links with "&" are not converted the same way as github links
On github markdown, a header like this one:
## This & That Section
will write out an id like this: `this--that-section`

On markdown-loader, it writes out `this-that-section`

This breaks links that adhere to github markdowns syntax:
```
[This & That Section](#this--that-section)
```

Another example:
```javascript
const marked = require("marked");

marked("# This & That Section");
// => '<h1 id="this-that-section">This &amp; That Section</h1>\n'`
```

Filed originally at https://github.com/peerigon/markdown-loader/issues/17

### LINKED ISSUE #1280: Duplicate Heading IDs
**Describe the bug**
In situations where a document has repeated headers, the ID is duplicated which is a big no-no for HTML validations (only one instance of an ID is allowed).

This is extremely common in changelogs where H3 headers are Added/Changed/Removed/Fixed etc. It flat out screws up accessibility as screen readers would likely ignore duplicate IDs and will cause formatting issues when an ID clashes with another outside the scope of the markdown render.

You cannot use IDs unless they are unique, so consider using "class" instead as a quick fix, and consider applying a configurable prefix to avoid clashing ('md-' perhaps).

**To Reproduce**
Create duplicate headers such as in [this demo](https://marked.js.org/demo/?outputType=html&text=%23%20test%20heading%0ASome%20words%0A%23%23%20Subheading%0ABlahblah%0A%23%23%20Subheading
). Then copy rendered HTML into [W3's HTML5 validato](https://validator.w3.org/nu/#textarea)r to see it fail validation

**Expected behavior**
No conflicts/render or screen reader issues.

### PR DESCRIPTION
**Pull request by styfle**
**Marked version:** 0.5.2

**Markdown flavor:** All

## Description

This fixes several bugs involving `headerIds: true`.

- Fixes #857
- Fixes #1280 
- Fixes #1354
- Fixes #1374

This adds a new export `marked.Slugger`.

I added several more tests to cover duplicates as well as ampersands.

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression
## Committer

In most cases, this should be a different person than the contributor.

- [x] Draft GitHub release notes have been updated.
- [x] CI is green (no forced merge required).
- [x] Merge PR

### CODE REVIEW

File: test/unit/marked-spec.js [hunk 1]
```diff
@@ -2,15 +2,29 @@ var marked = require('../../lib/marked.js');
...
-    var renderer = new marked.Renderer(marked.defaults);
-    var header = renderer.heading('test', 1, 'test');
-    expect(header).toBe('<h1 id="test">test</h1>\n');
+    var html = marked('# test');
+    expect(html).toBe('<h1 id="test">test</h1>\n');
+  });
+
+  it('should add unique id for repeating heading 1280', function() {
+    var html = marked('# test\n# test\n# test');
+    expect(html).toBe('<h1 id="test">test</h1>\n<h1 id="test-1">test</h1>\n<h1 id="test-2">test</h1>\n');
```
Comment (UziTech) [inline test/unit/marked-spec.js]: unit tests should only test a single function (unit) not the whole application.

We should expose `Slugger` on marked so we can write tests like:

```suggestion
    var slugger = new marked.Slugger();
    expect(slugger.slug('test')).toBe('test');
    expect(slugger.slug('test')).toBe('test-1');
    expect(slugger.slug('test')).toBe('test-2');
```

**Comment by UziTech** [comment]
we also need to test for https://github.com/Flet/github-slugger/pull/20

```js
    expect(slugger.slug('test 1')).toBe('test-1');
    expect(slugger.slug('test')).toBe('test');
    expect(slugger.slug('test')).toBe('test-2');
```

**Comment by styfle** [comment]
@davisjam @joshbruce After this PR lands, I can do a new release. Can you take a look?

**Comment by joshbruce** [comment]
Just to make sure I'm up to speed on the header conversation and whatnot. We're adding slugger to fix a defect, which is header duplication. We are still experiencing issues with unicode characters like Japanese and other glyph languages. And we are still progressing toward dropping header ID generation in the future. Yeah?

**Comment by styfle** [comment]
@joshbruce This PR fixes [non-latin char headings](https://github.com/styfle/marked/blob/632ac5d55d047e460c6c6ac70d81b8dc7920bc76/test/unit/marked-spec.js#L47) as well as duplicate IDs.

In the future we could drop support and rely on `github-slugger`. That PR #1354 (which was removing slugger from marked) was rejected. So I made this PR.

**Comment by tmorehouse** [comment]
We're receiving errors for this latest update:
```
8:57:27 PM: WARNING in ./src/components/table/README.md
8:57:27 PM: Module build failed (from ./node_modules/markdown-loader/index.js):
8:57:27 PM: TypeError: Cannot read property 'slug' of undefined
8:57:27 PM: Please report this to https://github.com/markedjs/marked.
8:57:27 PM:     at Renderer.heading (/opt/build/repo/node_modules/marked/lib/marked.js:962:17)
8:57:27 PM:     at Parser.tok (/opt/build/repo/node_modules/markdown-loader/node_modules/marked/lib/marked.js:1179:28)
8:57:27 PM:     at Parser.parse (/opt/build/repo/node_modules/markdown-loader/node_modules/marked/lib/marked.js:1130:17)
8:57:27 PM:     at Function.Parser.parse (/opt/build/repo/node_modules/markdown-loader/node_modules/marked/lib/marked.js:1112:17)
8:57:27 PM:     at marked (/opt/build/repo/node_modules/markdown-loader/node_modules/marked/lib/marked.js:1543:19)
8:57:27 PM:     at Object.module.exports (/opt/build/repo/node_modules/markdown-loader/index.js:14:12)
```

**Comment by tmorehouse** [comment]
We are adding a wrapper on the table renderer so that must be it. Will try restoring the slugger handler.

**Comment by tmorehouse** [comment]
It appears that `markdown-loader` is using `0.5.x`, which is causing our issue (as we are passing it a new renderer instance with code and table overrides)
